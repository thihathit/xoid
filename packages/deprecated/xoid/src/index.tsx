export type { Init, Atom, Listener, StateOf, GetState } from '@xoid/engine'
export type { Usable, Enhancer } from './core'

export { subscribe, effect } from '@xoid/engine'
export { create, use } from './core'
