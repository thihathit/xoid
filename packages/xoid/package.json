{
  "name": "xoid",
  "version": "0.9.27",
  "main": "./index.js",
  "module": "./esm/index.js",
  "types": "./index.d.ts",
  "sideEffects": false,
  "description": "Framework-agnostic state management library designed for simplicity and scalability",
  "homepage": "https://xoid.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/onurkerimov/xoid",
    "directory": "./packages/xoid"
  },
  "license": "MIT",
  "keywords": [
    "framework-agnostic",
    "state",
    "management",
    "state-management",
    "react",
    "redux",
    "recoil",
    "atom",
    "store"
  ],
  "dependencies": {
    "@xoid/engine": ">=0.9.25"
  }
}