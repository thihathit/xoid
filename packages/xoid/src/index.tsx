import type { Atom, Stream, Init, GetState, Usable } from './types'
import { createSelector, createApi, INTERNAL } from './internal/utils'
import { createEvent, createInternal } from './internal/lite'

export * from './types'

/**
 * Creates an atom with the first argument as the initial state.
 * Second argument can be used to attach more actions to the atom.
 * @see [xoid.dev/docs/api/create](https://xoid.dev/docs/api/create)
 */
export function create<T>(): Stream<T>
export function create<T>(init: Init<T>): Atom<T>
export function create<T>(init: Init<T>, getUsable: null): Atom<T>
export function create<T, U>(init: Init<T>, getUsable?: (atom: Atom<T>) => U): Atom<T> & Usable<U>
export function create<T, U = undefined>(
  init?: Init<T>,
  getUsable?: null | ((atom: Atom<T>) => U)
) {
  const isFunction = typeof init === 'function'
  const initialValue = (isFunction ? undefined : init) as T
  const internal = createInternal(initialValue, (() => (use as any).devtools.send(atom)) as any)
  if (isFunction) internal.get = createSelector(internal, init as (get: GetState) => T)
  const atom = createApi(internal)
  internal.usable = getUsable?.(atom)
  // devtools support
  Object.defineProperty(atom, 'debugValue', {
    set(debugValue: string) {
      internal.debugValue = debugValue
      ;(use as any).devtools.init(atom, initialValue, isFunction)
    },
  })
  return atom
}

/**
 * Gets the "usables" of an atom.
 * @see [xoid.dev/docs/api/use](https://xoid.dev/docs/api/use)
 */
export const use = <T extends any>(atom: Usable<T>): T =>
  (use as any).devtools.wrap((atom as any)[INTERNAL].usable, atom)
;(use as any).symbol = INTERNAL
;(use as any).createEvent = createEvent
;(use as any).devtools = {
  init: <T,>(_atom: T) => void 0,
  send: <T,>(_atom: T) => void 0,
  wrap: <T,>(value: T): T => value,
}
