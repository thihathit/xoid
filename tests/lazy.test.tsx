import { create, use, subscribe, Atom } from 'xoid'

const debug = <T,>(atom: Atom<T>) => {
  return {
    self: atom,
    selfSerialized: JSON.stringify(atom),
    get: atom(),
    getSerialized: JSON.stringify(atom()),
    use: use(atom as any),
  }
}

const consoleError = console.error
afterEach(() => {
  console.error = consoleError
})

it('lazily evaluates a state initializer function', () => {
  const fn = jest.fn()
  const atom = create(() => {
    fn()
    return 5
  })
  expect(fn).not.toBeCalled()
  expect(debug(atom)).toMatchSnapshot()
  expect(fn).toBeCalledTimes(1)
})

it('lazily evaluates a state initializer function 2', () => {
  const fn = jest.fn()
  const atom = create(() => {
    fn()
    return 5
  })
  expect(fn).not.toBeCalled()
  subscribe(atom, console.log)
  expect(fn).toBeCalledTimes(1)
})

it('lazily evaluate only when a sub atom is read/written', () => {
  const fn = jest.fn()

  const atom = create(() => {
    fn()
    return { deep: { value: 5 } }
  })
  expect(fn).not.toBeCalled()

  const subAtom = use(atom, (s) => s.deep.value)
  expect(fn).not.toBeCalled()

  subAtom(25)
  expect(fn).toBeCalledTimes(1)
})
