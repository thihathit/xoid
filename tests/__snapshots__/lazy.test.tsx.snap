// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazily evaluates a state initializer function 1`] = `
Object {
  "get": 5,
  "getSerialized": "5",
  "self": [Function],
  "selfSerialized": undefined,
  "use": undefined,
}
`;
